File: GO-DSA/singly_linked_list/push.go
package main

import "fmt"

type Node struct {
	val  string
	next *Node
}

type SingleLinkedList struct {
	head   *Node
	tail   *Node
	length int
}

func (s *SingleLinkedList) Push(val string) *SingleLinkedList {
	newNode := &Node{val: val}
	if s.head == nil {
		s.head = newNode
		s.tail = newNode
	} else {
		s.tail.next = newNode
		s.tail = newNode
	}

	s.length++
	s.PrintList()
	return s
}

func (s *SingleLinkedList) PrintList() {
	current := s.head
	for current != nil {
		fmt.Printf("%s -> ", current.val)
		current = current.next
	}
	fmt.Println("NULL")
}

func main() {
	list := &SingleLinkedList{}
	list.Push("Hello")
	list.Push("Goodbye")
	list.Push("!")
}

File: singly_linked_list/push.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class SinglyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  push(val) {
    var newNode = new Node(val);
    if (!this.head) {
      this.head = newNode;
      this.tail = this.head;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }
    this.length++;
    // console.log(list);
    this.printList();
    return this;
  }
  
  printList() {
    let current = this.head;
    let output = "";
    while (current) {
      output += current.val + " → ";
      current = current.next;
    }
    output += "NULL";
    console.log(output);
  }
}

var list = new SinglyLinkedList();
list.push("Hello");
list.push("Goodbye");
list.push("!");
